core/unix/rseq_linux.c

```diff
@@ -148,6 +163,118 @@ rseq_is_registered_for_current_thread(void)
+    /* We analyze the instructions inside [start,end) looking for register state used
+     * as inputs that we need to preserve for our restart.  We do not want to blindly
+     * spill and restore 16+ registers for every sequence (too much overhead), and we
+     * want a dead scratch register.
+     */
+    instr_t instr;
+    instr_init(GLOBAL_DCONTEXT, &instr);
+    app_pc pc = info->start;
+    int i;
+    bool reached_cti = false;
+    bool reg_written[DR_NUM_GPR_REGS];
+    memset(reg_written, 0, sizeof(reg_written));
+    memset(info->reg_used, 0, sizeof(info->reg_used));
+    info->scratch_reg = DR_REG_INVALID;
+    while (pc < info->end) {
+        instr_reset(GLOBAL_DCONTEXT, &instr);
+        app_pc next_pc = decode(GLOBAL_DCONTEXT, pc, &instr);
+        if (next_pc == NULL) {
+            REPORT_FATAL_ERROR_AND_EXIT(RSEQ_BEHAVIOR_UNSUPPORTED, 3,
+                                        get_application_name(), get_application_pid(),
+                                        "Rseq sequence contains invalid instructions");
+            ASSERT_NOT_REACHED();
+        }
+        if (instr_is_syscall(&instr)) {
+            REPORT_FATAL_ERROR_AND_EXIT(RSEQ_BEHAVIOR_UNSUPPORTED, 3,
+                                        get_application_name(), get_application_pid(),
+                                        "Rseq sequence contains a system call");
+            ASSERT_NOT_REACHED();
+        }
+        if (instr_is_call(&instr)) {
+            REPORT_FATAL_ERROR_AND_EXIT(RSEQ_BEHAVIOR_UNSUPPORTED, 3,
+                                        get_application_name(), get_application_pid(),
+                                        "Rseq sequence contains a call");
+            ASSERT_NOT_REACHED();
+        }
+        if (instr_is_cti(&instr))
+            reached_cti = true;
+        /* We only support GPR's as inputs.  We'd like to verify this, but it does
+         * not seem worth the effort (we would want to add new interfaces like
+         * DR_REG_START_SIMD or register iterators because we do not want to rule out all
+         * non-GPR use: only reads before writes).
+         */
+        for (i = 0; i < DR_NUM_GPR_REGS; i++) {
+            if (reg_written[i])
+                continue;
+            reg_id_t reg = DR_REG_START_GPR + (reg_id_t)i;
+            /* We do a quick and naive analysis.  We do not try to build a CFG, partly
+             * because core DR does not use such functionality *anywhere* (everything is
+             * deliberately linear), and partly because all bets are off with rseq
+             * anyway and it is not worth extra effort for corner cases.  We look for
+             * any read of a register prior to a write, completely ignoring control flow
+             * and treating the sequence as linear.  This is a search for inputs which
+             * should be read relatively early.  We document this limitation of only
+             * preserving clearly read-before-written register state.  We do not try to
```

Comment:
As the comments say, that is considered to be too much overhead.  16 stores and 16 loads is huge, far more work than the rseq sequence itself.  This is typically hot app code.  I don't like fragile handling of app code, but that ship sailed long ago with rseq where we already have a long list of things we can't handle, and I like big slowdowns even less.  I'm not super happy about anything related to rseq: none of it is perfect.

Reply:
Ok, just saying this is really an assumption that is easy to break, as soon as there is a little bit control flow in the rsequence. One had to run dominator analysis on it to fix. Maybe issue a warning if the rsequence has control flow?

Reply:
By the way, I didn't mean flat-rate save/restore all GPRs, just the ones that are actually used in the hopefully small sequence.

Reply:
If it's not too much overhead to scan the RSEQ section to find the GPRs that need to be save-restored, it might prevent future headaches as rseq becomes more prominent in open source software.

Otherwise, I am still in the camp of "if we have assumptions, let's try to verify them rather than actually assume them". So I would vote for an assert where reasonable to check assumptions; a TODO would be fine too, if it's too much to verify this assumption in this PR.

Reply:
Just in case, the question is to decide between

a) fully blown live-in (reachability) analysis of the real control flow to find the real live-ins of the sequence
b) assume linear control flow (is it too fragile or ok?)
c) conservatively assume all uses are live-in regardless whether they are (how man are there usually, is it too much overhead?)

Reply:
I thought about preserving all registers read anwyhere, but it is twice as many.  Computing the average across 14 examples:

Everything: 16
Read-anywhere: 5.7
Read-linear: 2.8
Build CFG and look for what's clobbered too: 1.0

Clearly building the CFG and doing the full analysis is best, but it will be the most effort and I don't know that the code can be leveraged for anything else in DR, as the comments in this PR say.  It does feel the best: I don't really like the other solutions.

However, consider complicating factors: indirect branches inside the region (I think my current prototype for running a local copy requires that indirect branches inside the region exit the region: that would help here too), and callouts to helper functions outside the region: we would have to analyze the helper too, and it could run arbitrary amounts of code.  Some of our examples used to call to helpers.  Since our main use cases no longer use helpers I added a ban on them here, but it's just another place where there are so many complications in trying to handle rseq, and we are forced to have a long list of assumptions and limitations.  It is ugly and  not fun to deal with.

Reply:
As discussed offline, we decided on a better approach: save the registers that are *written* anywhere in the rseq sequence, regardless of control flow.  This also handles cases where a register is live across the entire rseq sequence and read *afterward*, but with one branch inside the sequence which clobbers it.